### ISSUE TITLE
SQL value more efficient as bytes

### ISSUE THREAD

**Issue by ricardofandrade** [gofrs/uuid#48]
The SQL [`Value()`](https://github.com/gofrs/uuid/blob/d41eeda0759468834c84365a32d1b1ec4264c75f/sql.go#L32) gives a string representation of UUID when called. However, storing UUIDs as bytes may be more efficient for some [databases](http://mysqlserverteam.com/storing-uuid-values-in-mysql-tables/).
Other UUID libraries have the same issue: https://github.com/google/uuid/issues/20

Would it be possible to have an option to either set globally to output bytes or a similar approach used for `NullUUID`?

I'd be glad to work on a patch in either direction if necessary.

**Comment by zerkms** [comment]
I think one possible implementation would be to create a wrapper type:

1. Whose constructor accepts `uuid.UUID`
2. Which implements `driver.Valuer` that returns `[]byte`

In that case it does not incur a BC, and follows the common sense design practices.

A quick and dirty API sample (with deliberately verbose declarations):

```go
var u UUID = ...

sqlByteUuid := NewSQLByteUUID(u)

var b []byte
var err error

b, err = sqlByteUuid.Value()
```

**Comment by ricardofandrade** [comment]
@zerkms thanks for the quick reply.
Since two SQL type conversions are already included (`UUID` and `NullUUID`), would you propose this wrapper type to also be included in `uuid`?

**Comment by ricardofandrade** [comment]
I am currently experimenting with your design.I may propose a change if the ergonomics of using it within `uuid` are better.

**Comment by theckman** [comment]
I think it's solvable in a backwards compatible way without _another_ type.

**Comment by ricardofandrade** [comment]
Are you proposing extending `NullUUID` @theckman or `UUID` itself?

**Comment by theckman** [comment]
I'm naively thinking of extending the NullUUID type to have a mode, which allows you to use either a string or raw bytes mode. GopherCon has me a little distracted, so I'll admit I'm to the level of spitballing while not having thought it through fully. 😊

The one part I've not thought of is pulling the data from the DB. We could maybe use the data length to know if it's a set of bytes, or a string itself.

**Comment by ricardofandrade** [comment]
When pulling from the database what @theckman  said is already being done:
https://github.com/gofrs/uuid/blob/aeac6c4a7786adf624dc27d38504c9011b1ffd58/sql.go#L40

The problem is that when you write back to the database only string is returned:
https://github.com/gofrs/uuid/blob/aeac6c4a7786adf624dc27d38504c9011b1ffd58/sql.go#L33

This inconsistency needs to be adjusted on the layer of code between the SQL driver and UUID which ends up polluting the application logic in one way or the other.

Ideally, `Value()` should receive a hint of the format from the SQL driver so that the implementation of value could decide the output.

Since that's not going to change unfortunately, we have a few options:
- A runtime setting that switches the format globally: hum, ok, no.
- A type that always return `[]byte`: that doesn't work when you may switch over the two formats during runtime (my case). So this rules out the `SQLByteUUID` suggestion.
- A type that let's you decide the output format: this discussion.

Now, if add yet another type, let's call it `SQLUUID` you will also a `NullSQLUUID` given that embedding may be discouraged.

So I am inclined to agree with @theckman that maybe extending `NullUUID` may be an option.
For example:
```go
type NullUUID struct {
	UUID  UUID
	Valid bool
	Binary bool
}

func NewBinaryNullUUID(u UUID, valid bool) {
	return NullUUID{ UUID: u, valid, true }
}
```

My only concerns would be that:
- `NullUUID` doesn't imply support to the binary format.
- `NewBinaryNullUUID` would be just one of many constructors needed

P.S.: Please note also that all these solutions doesn't fix the fact that `gofrs/uuid` automatically detects the format on database retrieval but expects the format to be set on store.

**Comment by dylan-bourque** [comment]
In this scenario, there's no context passed into `Value()` that would allow `NullUUID` to make a runtime determination of what the correct output format (binary/string) should be.  Also, many database drivers are _smart enough_ to accept the string representation from the code and convert it to the DBMS's internal format.  For sure, both PostgreSQL and SQL Server do the right thing.

On the surface, extending `NullUUID` like this feels like it would complicate the type's interface unnecessarily.

**Comment by acln0** [comment]
> So I am inclined to agree with @theckman that maybe extending NullUUID may be an option.
> For example:
> 
> ```
> type NullUUID struct {
> 	UUID  UUID
> 	Valid bool
> 	Binary bool
> }
> 
> func NewBinaryNullUUID(u UUID, valid bool) {
> 	return NullUUID{ UUID: u, valid, true }
> }
> ```

I'm not convinced by this design. The `Binary` field is only meaningful for UUIDs that are about to be encoded and sent out to a database. By contrast, the `Valid` field is meaningful in both directions. There is indeed no good way to provide more context to `Value()` besides adding a field on the receiver, but I'm not convinced this is the path we should be taking.

> Also, many database drivers are smart enough to accept the string representation from the code and convert it to the DBMS's internal format.

I know little about this problem space, but if this is true, then it represents even more reason to take no action here.

The design @zerkms proposed seems simple and elegant to me. Callers who are interested in this particular optimization (and who don't get help from the database, as described above) can introduce the wrapper type in their own scope, at the boundary between their business logic and the database persistence code. I don't think such a type belongs in the `uuid` package itself.

My opinion is that we should take no action.

**Comment by ricardofandrade** [comment]
For most SQL drivers I've worked with in different languages the knowledge about the type conversion would go both ways. Like I said it's unfortunate `Value()` doesn't have this context.

@dylan-bourque you're right about PostgreSQL _when the driver is operating in the text format_.
I have not tried the go driver operating in the binary format though.

However, SQLite is transparent and let you store whatever format you want.
Same goes for MySQL which allows using `BINARY(16)` for UUID.
Both widely enough used databases to worth this discussion, I believe.

Please note there's nothing preventing both formats to be used _at the same time_ in SQLite and MySQL!
I mean, the same struct containing `UUID` being serialized as binary in one place and as text another.
Not the most beautiful design but possible. I personally have to deal with different databases (therefore different formats) for the same given struct. 

In general, I'd agree with @dylan-bourque about complicating `NullUUID` interface, on the other hand:
* It's intended to be used with databases (otherwise it wouldn't be in `sql.go`), binary output could be considered yet another step towards a wide support for different databases.
* However adopts this type has already the expectation of a more complicated interface.
* Having a fixed `BinaryUUID` won't work because a type that always output binary doesn't allow support two distinct databases using respectively text and binary representations at the same time
* Having two other types `MaybeBinaryUUID` and `NullMaybeBinaryUUID` will certainly be confusing to explain.

**Comment by ricardofandrade** [comment]
Some food for thought...
Given:
```sql
-- MySQL schema
CREATE TABLE data(id BINARY(16));
```
And:
```sql
-- PostgreSQL schema
CREATE TABLE data(id UUID);
```
We have:
```go
type Data struct {
    Id uuid.UUID `db:"id"`
}

pgDb := getPostgreSQLConnection()
myDb := getMySQLConnect()

id, _ := uuid.NewV4()
data := Data{ Id: id }
pgDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // works
myDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // fails - string(36) is not byte(16)
```

**Comment by dylan-bourque** [comment]
Good point about `BINARY` database columns, @ricardofandrade.  I honestly wasn't considering that before.  We actually had several folks together at GopherCon today and discussed it some more.

If we make a change, it would be close to your proposal where the caller would have to opt-in to having `Value()` return bytes instead of a string.

**Comment by ricardofandrade** [comment]
TL; DR - The idea of a separate type for databases expecting a binary UUID representation shows more shortcomings the more I think about this whole thing. So does the idea of extending `NullUUID`. This would be solvable in the database driver if there were one true UUID implementation but since there isn't one I propose adding to this package a global control for the output format of `UUID` itself as a more sane way to deal with databases using the binary UUID representation.

Here are the different paths I followed and what led my to settle with the proposal above.

### Solution 1 - Separate type

Let's start with the most elegant solution, a completely separate type that always output binary. Let's call it `BinaryUUID`. My example above would look like this:
```go
// uuid code
type BinaryUUID struct {
    UUID
}

func (b BinaryUUID) Value() (driver.Value, error) {
    return b.Bytes(), nil
}

// user code

type Data struct {
    Id uuid.UUID // No database representation
}

type PgData struct {
    Id uuid.UUID `db:"id"`
}

type MyData struct {
    Id uuid.UUID `db:"id"`
}

pgDb := getPostgreSQLConnection()
myDb := getMySQLConnect()

id, _ := uuid.NewV4()
data := Data{ Id: id }

pgData := data
pgDb.Exec("INSERT INTO data(id) VALUES (?)", &pgData) // works

myData := BinaryUUID(data) 
myDb.Exec("INSERT INTO data(id) VALUES (?)", &myData) // works
```

Advantages:
* It's clean and idiomatic to go.

Problems:
* The user have to pay the penalty of a conversion if their DB uses binary UUIDs.
* Or (unexpectedly) `BinaryUUID` will be littering the entire code base replacing `UUID`.
* If multiple databases are used the user must implement variants of their own types for text and binary.
* If `NULL` is required then another `NullBinaryUUID` is needed.
* The documentation needs to be very clear about when to use each type.

> Note: The way the conversion to `BinaryUUID` is done could vary (i.e. constructor) but that doesn't change the fact it needs to be done manually.

### Solution 2 - Extending NullUUID

The second option offered was adding a `Binary` flag on `NullUUID` so it would be the de-facto database type with all the bells and whistles needed to work in a broad range of situations. My example above would look like this with it:
```go
// uuid code
type NullUUID struct {
    UUID UUID
    Valid bool
    Binary bool
}

func (n NullUUID) Value() (driver.Value, error) {
    if !b.Valid { return nil, nil }
    if b.Binary { return n.Bytes(), nil }
    return n.String(), nil
}

// user code
type Data struct {
    Id uuid.NullUUID `db:"id"`
}

pgDb := getPostgreSQLConnection()
myDb := getMySQLConnect()

id, _ := uuid.NewV4()
data := Data{ Id: id }

pgDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // works

data.Binary = true
myDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // works
```

Advantages:
* Visually, a more straightforward use when compared to different types.
* Because of the single type, the user doesn't need to have variants of their own types for different DBs.

Problems:
* Again here, the user have to pay a penalty to indicate the format if their DB uses binary for every single use.
* The fact that a single type is used makes indicating the format a runtime thing, repetitive and error prone, done for every single SQL statement storing a UUID to the database.
* `NullUUID` will be confusing and a problem when the database is schema is `NOT NULL`.
* As some pointed out already, the interface of `NullUUID` gets more complex and even though `Valid` and `Binary` wouldn't interact with each other some may thing otherwise and get confused.
* The user does not immediately know that `NullUUID` is the thing to be used for binary representation. In that sense a separate type make that much more obvious.

> Note: How the `Binary` flag is set can vary in implementation but it doesn't change the fact it needs to be set before each individual SQL statement dealing with the binary UUID.

### Solution 3 - A separate type with both NULL and binary support

Let's call this `SQLUUID`. It's the same as the extended `NullUUID` but with a different name. The different name would (and should) make it obvious it's a type intended to support a broad range of databases. Then, documenting the binary support would be easier.
However, all the other problems found in the extended `NullUUID` would still be present, specially being error prone to use.

### Solution 4 - A global option

This was my initial idea. I didn't like it and had ditched it completely once others offered more elegant suggestions. But after carefully considering all the other solutions I had to give it another shot.
My example would look like this with it:
```go
// uuid code
var binaryFormat = false
func ForceSQLBinary() {
    binaryFormat = true
}

func ForceSQLText() {
    binaryFormat = false
}

func (u UUID) Value() (driver.Value, error) {
    if binaryFormat { return n.Bytes(), nil }
    return n.String(), nil
}

func (n NullUUID) Value() (driver.Value, error) {
    if !b.Valid { return nil, nil }
    if binaryFormat { return n.Bytes(), nil }
    return n.String(), nil
}

// user code
type Data struct {
    Id uuid.NullUUID `db:"id"`
}

id, _ := uuid.NewV4()
data := Data{ Id: id }

pgDb := getPostgreSQLConnection()
uuid.ForceSQLText()
pgDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // works

myDb := getMySQLConnect()
uuid.ForceSQLBinary()
myDb.Exec("INSERT INTO data(id) VALUES (?)", &data) // works
```

Advantages:
* As for other solutions, this is also backwards compatible.
* As for the solutions 2 & 3, this doesn't need separate user types for different databases.
* As for the solutions 2 & 3, the user has to manually indicate which format the database is expecting but as you may have noticed in the way I reordered things in the example usually the user *just needs to do that once*, at the time the database flavor is known and it requires binary UUID.

One small problem:
* Since it's global, the calls to `ForceSQLBinary()` would be far away from where the UUID gets written to the database affecting maintainability. However, it's much less user code than the alternatives which I'd say neutralizes this problem.

One big problem:
* If the user where to switch (*during runtime*) among databases supporting different formats it certainly would cause contention because of the global state. Which is bad as solutions 2 & 3 (needs to be called before each SQL statement) plus a mutex around each call to either `ForceSQLText()` and `ForceSQLBinary()`.

> Note: please remember the above is only a problem if the application is writing to two different databases using different UUID representations *concurrently*. Reading from the database is not a problem because `uuid.Scan(interface{})` deals with both `[]byte` and `string`. 

Though, most people don't run into this problem and when they do it's most likely they would be willing to pay the cost of a more complex solution that doesn't involve calling `ForceSQLBinary()` at all.

Let's examine this assumption:
1. Single database use case (UUID is text): nothing needs to be done.
2. Single database use case (UUID is binary): a single call to `ForceSQLBinary()` during startup.
3. Multi-database use case (UUID is always text): same as 1.
4. Multi-database use case (UUID is always binary): same as 2.
5. Multi-database use case (UUID format depends on the driver): detect the driver used and call `ForceSQLBinary()` once if needed.
6. Multi-database use case (migration from text to binary or vice-versa): data is exported from one database and imported into another. Given the serial nature of the operation a pair of calls to `ForceSQLBinary()` and `ForceSQLText()` happens for each batch that needs conversion.
7. Multi-database use case (both binary and text UUID at the same time, i.e. writing concurrently to multiple databases): the "big problem" mentioned above.

Global state may not be nice in general but from the practical point of view this would be the solution requiring the least amount of changes from this UUID implementation (less things to maintain) and the easiest one to be used.

Documenting the limitations of the global state under concurrency should be sufficient to warn users.

### Solution 5 - This is a database driver problem, let's drivers fix it

Since it's the database that requires a different format (binary) then the driver for that database should solve this problem, right? Not quite.

A few problems:
* User-defined: Like mentioned before both SQLite and MySQL leave the format to be defined by the user either in the schema (`BINARY`/`BLOB`) or when data gets written (SQLite).
* No standard UUID: If there were a single UUID package and it were to be part of the standard library, all drivers could rely on its existence to decide on the representation, like it's done for `time.Time` and `time.Duration`. However, this is not the case and it's unlikely drivers will support a handful of different UUID implementations.
* Lack of context: The `sql` package implementation gives no context to the driver about the schema at the point the data gets written to the database. Different from when the data is retrieved from the database there's no notion of column types. So it's impossible for a driver to decide about the storage format for any given type. This information must come from the application.

All said, it's not feasible to ask driver implementers to deal with the problem of the binary UUID representation.

### Solution 6 - Do nothing

Here's what I need to do if nothing is done to support the binary UUID representation:
* SQLite: I am injecting a go function (let's call it `adopt_uuid()`) that parses the string representation returned by `uuid.UUID.Value()` and calls `uuid.UUID.Bytes()`.
* MySQL: I created an `adopt_uuid()` function that does the same conversion from the string representation to binary.
* PostgreSQL: I created no-op function with the same name used for the other databases so I can have a similar SQL most of the time.

The SQL is littered with conversion functions (and the cost associated to them) whenever I am writing a UUID. Others might roll out their own versions of the solutions listed here.

## Conclusion

If you have got this far I hope it has been an informative journey convincing you that:
* Adding new types doesn't cover some important use cases and the ergonomics of using those are not great giving the problem at hand. They probably would just cause confusion and result in more code to maintain.
* Switching globally the output format between binary and text is a reasonable compromise with a known problem I expect people to easily identify and avoid.
* Defer to driver implementers is not feasible unfortunately.
* Doing nothing impacts users of this package negatively which then need to rely on rolling their own types or use "creative" SQL.

**Comment by acln0** [comment]
Thank you for the proposal and the detailed write-up.

---

> If the user where to switch (during runtime) among databases supporting different formats it certainly would cause contention because of the global state. Which is bad as solutions 2 & 3 (needs to be called before each SQL statement) plus a mutex around each call to either `ForceSQLText()` and `ForceSQLBinary()`.

This isn't the only big problem. Consider the following diagram, where boxes represent packages, and arrows represent dependencies.

```
             +-----+
         +---+  X  +---+
         |   +-----+   |
         |             |
         |             |
      +--v--+       +--v--+
   +--+  Y  +-+   +-+  Z  +--+
   |  +-----+ |   | +-----+  |
   |          |   |          |
   |          |   |          |
+--v--+     +-v---v-+     +--v--+
|mysql|     |  uuid |     | pg  |
+-----+     +-------+     +-----+
```

Consider what happens if `Y` calls `uuid.ForceSQLBinary` in its `init` function, and `Z` calls `uuid.ForceSQLText` in its `init`. If the author of `X` does not control `Y` and `Z` as well, they are entirely helpless, and cannot use `Y` and `Z` together. The observed behavior depends on the initialization order of the packages, and is broken either way.

From their own perspective, the authors of `Y` and `Z` did nothing wrong. They just used the available API of package `uuid`. I argue that package `uuid` should be stateless, and not allow such situations to occur. If the authors of `Y` or `Z` want to change the way UUIDs are interpreted by the SQL driver they use, they must do so in their own scope, through whatever means they see fit.

---

> Though, most people don't run into this problem and when they do it's most likely they would be willing to pay the cost of a more complex solution [...]

From where we stand now, I think it's possible to make the same argument for the current situation, without introducing any additional API and shifting the complexity elsewhere.

---

> Doing nothing impacts users of this package negatively which then need to rely on rolling their own types or use "creative" SQL.

It seems to me like the simplest change callers can make, in order to use the binary form, rather than the text form, is to pass `u[:]` as the argument to queries, rather than `u`. Schematically, given a `var u uuid.UUID`,
```
db.Exec(`INSERT INTO some_table (id, name) VALUES (?, ?)`, u[:], "test")
```

The complete test program I ran against sqlite is available [here](https://play.golang.org/p/ezKNk6ixwwD). I haven't tried it for MySQL, but I'd be very surprised if it didn't work. In my opinion, this isn't prohibitively difficult, nor does it impose a significant complication of client code.

As discussed already, specialized types scoped to the caller's database access layer can also work, and perhaps some users would be more comfortable with that.

One could argue that it's easy to forget the `[:]`, or that it's easy to forget to wrap the UUID in the specialized type, but I think the same argument can be made for mistakes like forgetting an argument to a query entirely, or reversing the order of two arguments.

---

> Documenting the limitations of the global state under concurrency should be sufficient to warn users.

I think documenting the current behavior of `(UUID).Value` could serve a similar purpose. We could document that `Value` returns the string form, and that if callers need the binary form, they should pass `u[:]` into queries.

---

> Global state may not be nice in general but from the practical point of view this would be the solution requiring the least amount of changes from this UUID implementation (less things to maintain) and the easiest one to be used.

It may be the easiest one to be used for many cases -- perhaps even for most cases -- but I cannot consider it a reasonable (or even acceptable) solution if it introduces the possibility for catastrophic diamonds like the one described above to occur.

---

All this being said, I am currently leaning pretty strongly towards a "no" for this proposal, sorry. In the more general sense, I think this is a problem that should be solved in client code, because I don't see enough evidence and a compelling enough design to suggest otherwise.

**Comment by ricardofandrade** [comment]
@acln0 thanks for the reply.

I ignored the package dependency aspect because I don't think other packages would themselves call `ForceSQLBinary()` or `ForceSQLText()`. Those are application level decisions which requires deep knowledge of *your* database schema in order to be called. However, there's also no way to restrict these calls to the `main` package which puts a maintainer of `uuid` in a uncomfortable situation, I will have to agree with you on that.

Your suggestion of using `u[:]` ignores important aspects I considered in my proposal detailed below.

#### Tagged Structs

My examples are all purposely based on tagged structs where the user doesn't need to refer to fields individually. It means there's simply no place to do `u[:]` because the `u` is never passed by the user, it's not application code but part of the `sql` package internals.

Your suggestion requires this:
```go
type Data struct {
    Id uuid.NullUUID `db:"id"`
}
id, _ := uuid.NewV4()
data := Data{ Id: id }
myDb := getMySQLConnect()
myDb.Exec("INSERT INTO data(id) VALUES (?)", &data)
```

To be done this way:
```go
myDb.Exec("INSERT INTO data(id) VALUES (?)", data.Id[:])
```

Which now requires touching this code every time `Data` gets fields added or removed.

#### Multiple databases, same SQL

Your idea would only be valid if the code would run always in the same database and that particular database only does binary UUID. Simple CRUD SQL statements are generally accepted by most databases without changes. Unfortunately your change requires a different SQL when the database uses a binary UUID.

#### Penalization of databases using binary UUID but only for writing

Finally, your idea makes things exponentially more difficult for writing binary UUIDs. You're implying that tagged structs, aiming for compatible SQL and supporting multiple databases is out of reach if a binary UUID is throw in the mix. Thing that are perfectly fine with textual UUID.
And don't forget that reading binary UUID is perfectly fine.

#### Closing thoughts

I'll have to concede for now that none of the solutions proposed so far are worth to be introduced permanently as part of `uuid` and that maybe only enhancements to `sql` would make any change in `uuid` worthwhile. I'll keep digging and I'll be glad to bring any findings here again.

I hope the info here serve to help others with similar problems. 😃

**Comment by rbastic** [comment]
Hi,

Going to close this now since nothing is actionable. Would like to avoid the chance that passing potential or active users see this and become mired in a long "issue", thinking it's a bug/problem.

Thanks again for your research, questions, and information!

Additionally, if possible please bring performance discussion with this issue to slack #gofrs before reopening.

Warm regards,
